### PR TITLE
DenseNet Implementaion

### PR DESCRIPTION
**Pull request by Andr0id100**
Any feedback would be awesome.
@rxwei @BradLarson

### CODE REVIEW

**Comment by Andr0id100** [comment]
Is that alright @rxwei ?

File: Models/ImageClassification/DenseNet.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// limitations under the License.
+
+import TensorFlow
+
+// Original Paper:
+// Densely Connected Convolutional Networks
+// Gao Huang, Zhuang Liu, Laurens van der Maaten, Kilian Q. Weinberger
+// https://arxiv.org/pdf/1608.06993.pdf
+
+public struct Conv: Layer {
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: This public struct should not be defined at top level because it's very specific to this model. When defined at top level, all users importing `ImageClassification` would get a public `Conv`, which is a leaky abstraction that would cause confusion. This has been sub-optimal in other `ImageClassification` models too, e.g. the public [`BatchNormConv2DBlock `](https://github.com/tensorflow/swift-models/blob/a60989d48bce41066de33dc2af280813b54232e1/Models/ImageClassification/WideResNet.swift#L23) struct.

Let's think about what's the right solution moving forward. I'd suggest doing a major refactoring to put any auxiliary layers in an extension of the primary model type. So this model could look like:

```swift
public struct DenseNet: Layer {
    public var conv = Conv(...)
    public var maxpool = MaxPool2D<Float>(...)
    ...
}

extension DenseNet {
    public struct Conv: Layer {
        ...
    }
}
```

What do you think about this approach? If we reach consensus on this, we can start a PR to refactor all other models this way.

Comment (Andr0id100) [reply Models/ImageClassification/DenseNet.swift]: I think that would be a good idea.

Comment (Andr0id100) [reply Models/ImageClassification/DenseNet.swift]: Does something like this looks good?

**Comment by Andr0id100** [comment]
Does something like this looks good?

File: Models/ImageClassification/DenseNet.swift [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            )
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            return conv(relu(bNorm(input)))
+        }
+    }
+
+    // Seperate structure for the 1x1 and 3x3 conv layers whose collection
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: For a documentation comment, use a formal sentence to describe `ConvPair`. I'd suggest something like the following:

_A pair of a 1x1 `Conv` layer and a 3x3 `Conv` layer._

File: Models/ImageClassification/DenseNet.swift [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            let conv3Output = conv3x3(conv1Output)
+            return conv3Output.concatenated(with: input, alongAxis: -1)
+        }
+    }
+
+    public struct DenseBlock: Layer {
+        public var pairs: [ConvPair] = []
+
+        public init(repitionCount: Int, growthRate: Int = 32, inputFilterCount: Int) {
+            for i in 0 ..< repitionCount {
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]:
```suggestion
            for i in 0..<repetitionCount {
```

File: Models/ImageClassification/DenseNet.swift [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                let filterCount = inputFilterCount + i * growthRate
+                pairs.append(ConvPair(inputFilterCount: filterCount, growthRate: growthRate))
+            }
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            let pairsReduced = pairs.differentiableReduce(input) { last, layer in
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: Return this directly.
```suggestion
            pairs.differentiableReduce(input) { last, layer in
```

File: Models/ImageClassification/DenseNet.swift [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                layer(last)
+            }
+            return pairsReduced
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]:
```suggestion
```

File: Models/ImageClassification/DenseNet.swift [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                filterSize: 1,
+                inputFilterCount: inputFilterCount,
+                outputFilterCount: inputFilterCount / 2
+            )
+            pool = AvgPool2D(poolSize: (2, 2), strides: (2, 2), padding: .same)
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            return input.sequenced(through: conv, pool)
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]:
```suggestion
            input.sequenced(through: conv, pool)
```
We can omit the `return` in functions whose body has a single expression.

File: Models/ImageClassification/DenseNet.swift [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            conv = Conv2D(
+                filterShape: (filterSize, filterSize, inputFilterCount, outputFilterCount),
+                strides: (stride, stride),
+                padding: .same
+            )
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            return conv(relu(bNorm(input)))
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]:
```suggestion
            conv(relu(bNorm(input)))
```

File: Models/ImageClassification/DenseNet.swift [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            let conv1Output = conv1x1(input)
+            let conv3Output = conv3x3(conv1Output)
+            return conv3Output.concatenated(with: input, alongAxis: -1)
+        }
+    }
+
+    public struct DenseBlock: Layer {
+        public var pairs: [ConvPair] = []
+
+        public init(repitionCount: Int, growthRate: Int = 32, inputFilterCount: Int) {
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: Typo: "repition" -> "repetition"

File: Models/ImageClassification/DenseNet.swift [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            return conv(relu(bNorm(input)))
+        }
+    }
+
+    /// A pair of a 1x1 `Conv` layer and a 3x3 `Conv` layer.
+    // make a dense block
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]:
```suggestion
```

File: Models/ImageClassification/DenseNet.swift [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        let level2 = level1.sequenced(through: denseBlock2, transitionLayer2)
+        let level3 = level2.sequenced(through: denseBlock3, transitionLayer3)
+        let output = level3.sequenced(through: denseBlock4, globalAvgPool, dense)
+        return output
+    }
+}
+
+extension DenseNet {
+    public struct Conv: Layer {
+        public var bNorm: BatchNorm<Float>
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: Avoid abbreviations, especially single-character ones.
```suggestion
        public var batchNorm: BatchNorm<Float>
```

**Comment by Andr0id100** [comment]
Thanks for all the feedback.

File: Models/ImageClassification/DenseNet.swift [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            )
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            conv(relu(batchNorm(input)))
+        }
+    }
+
+    // A pair of a 1x1 `Conv` layer and a 3x3 `Conv` layer.
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: Please use three slashes for documentation comments.

File: Models/ImageClassification/DenseNet.swift [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        public init(repetitionCount: Int, growthRate: Int = 32, inputFilterCount: Int) {
+            for i in 0..<repetitionCount {
+                let filterCount = inputFilterCount + i * growthRate
+                pairs.append(ConvPair(inputFilterCount: filterCount, growthRate: growthRate))
+            }
+        }
+
+        @differentiable
+        public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
+            return pairs.differentiableReduce(input) { last, layer in
```
Comment (rxwei) [inline Models/ImageClassification/DenseNet.swift]: Omit the ‘return’ keyword for consistency.

**Comment by Andr0id100** [comment]
I am thinking of giving it a structure similar to the ResNet file to incorporate all the different types.
I could change the name to DenseNet121 for now.

File: Tests/ImageClassificationTests/Inference.swift [hunk 13]
```diff
@@ -21,6 +21,15 @@ final class ImageClassificationInferenceTests: XCTestCase {
     override class func setUp() {
         Context.local.learningPhase = .inference
     }
+    
+    func testDenseNet() {
```
Comment (rxwei) [inline Tests/ImageClassificationTests/Inference.swift]:
```suggestion
    func testDenseNet121() {
```

**Comment by Andr0id100** [comment]
I think this can be merged now.
